src/hotspot/share/utilities/globalDefinitions_gcc.hpp

```diff
@@ -138,22 +138,18 @@ inline int g_isfinite(jdouble f)                 { return isfinite(f); }
 #define FORMAT64_MODIFIER "ll"
 #endif // _LP64
 
-// gcc warns about applying offsetof() to non-POD object or calculating
-// offset directly when base address is NULL. The -Wno-invalid-offsetof
-// option could be used to suppress this warning, but we instead just
-// avoid the use of offsetof().
-//
-// FIXME: This macro is complex and rather arcane. Perhaps we should
-// use offsetof() instead, with the invalid-offsetof warning
-// temporarily disabled.
-#define offset_of(klass,field)                          \
-[]() {                                                  \
-  char space[sizeof (klass)] ATTRIBUTE_ALIGNED(16);     \
+// GCC/Clang warns about applying offsetof() to types that are not standard layout or calculating
+// offset directly when base address is NULL. Technically using offsetof() on non-standard layout is
+// undefined behavior before C++17 and "conditionally-supported" after C++17. Rather than take our
+// chances we use our own implementation which both GCC/Clang ultimately produce a constant for but
+// is not a constant expression.
+#define offset_of(klass, field)                         \
+([]() __attribute__((always_inline)) {                  \
+  alignas(klass) char space[sizeof(klass)];             \
   klass* dummyObj = (klass*)space;                      \
   char* c = (char*)(void*)&dummyObj->field;             \
   return (size_t)(c - space);                           \
-}()
-
+}())
```

Comment:
The xlc version of this macro is the same as the old (before JDK-8294902) one from here, with a
comment saying it is the same as the gcc version.  Clearly that's no longer true.  Either that one
should also be updated or the comment updated.  That's probably something for the aix-ppc port
maintainers to decide.  There ought to be a followup bug report for it though.